packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx

```diff
@@ -140,30 +141,51 @@ const StartVisitForm: React.FC<DefaultWorkspaceProps> = ({ patientUuid, closeWor
     formState: { errors },
   } = methods;
 
+  const createPayload = (data: VisitFormData) => {
+    const { timeFormat, visitDate, visitLocation, visitTime, visitType, visitAttributes } = data;
+    const [hours, minutes] = convertTime12to24(visitTime, timeFormat);
+
+    return {
+      patient: patientUuid,
+      startDatetime: toDateObjectStrict(
+        toOmrsIsoString(
+          new Date(dayjs(visitDate).year(), dayjs(visitDate).month(), dayjs(visitDate).date(), hours, minutes),
+        ),
+      ),
+      visitType: visitType,
+      location: visitLocation?.uuid,
+      attributes: Object.entries(visitAttributes)
+        .filter(([key, value]) => !!value)
+        .map(([key, value]) => ({
+          attributeType: key,
+          value: value as string,
+        })),
+    };
+  };
+
+  const handleRetrospectiveVisit = useCallback((data: VisitFormData, event) => {
+    const { visitDate } = data;
+    const payload = createPayload(data);
+
+    const dispose = showModal('retrospective-modal', {
+      closeModal: () => {
+        dispose();
+        closeWorkspace();
+      },
+      onCancel: () => dispose(),
+      patientUuid,
+      visitDate,
+      payload,
+    });
+  }, []);
+
   const onSubmit = useCallback(
     (data: VisitFormData, event) => {
-      const { timeFormat, visitDate, visitLocation, visitTime, visitType, visitAttributes } = data;
-
-      setIsSubmitting(true);
+      const { visitDate, visitLocation } = data;
 
-      const [hours, minutes] = convertTime12to24(visitTime, timeFormat);
+      const payload = createPayload(data);
```

Comment:
We can use `useMemo` here

Reply:
We're already in a `useCallback()`... I think this is fine.